### DISCUSSION TITLE
How to implement custom XCOM

### DISCUSSION THREAD

**Discussion by waardd** [apache/airflow#29059] [Q&A]
I want to use a custom XCOM to be able to share variables between different tasks.
I have found a custom XCOM like so:

import uuid
import json
from airflow.models.xcom import BaseXCom

PREFIX = 'XCOM_'

class CustomXcomFS(BaseXCom):
    """Example Custom Xcom persistence class."""

    @staticmethod
    def serialize_value(value):
        key_str = PREFIX + str(uuid.uuid4())
        with open("/tmp/"+key_str, 'w') as outfile:
            json.dump(value, outfile)
        print("Xcom with key-",key_str, " written to filesystem")
        return BaseXCom.serialize_value(key_str)

    @staticmethod
    def deserialize_value(result):
        result = BaseXCom.deserialize_value(result)
        # Check if result is string and have XCOM as prefix
        if isinstance(result, str) and result.startswitch(PREFIX):
            with open(f'/tmp/{result}', 'r') as jsonfile:
                data = json.load(jsonfile)
                return data
        return result
    
    def orm_deserialize_value(self):
        return 'XCOM writen to filesystem'
        
 I know you have to change the XCOM backend setting to set the location
 For example airflow_xcom_backend: "library.custom_xcom_backend.CustomXcomFS" ( <- in my ansible book )
 
 But if i run this, will it replace my XCOM location to only /tmp/ (on a linux system) or will both be in place (so the task_ids XCOM and the filesytem /tmp) ??

**Comment by potiuk** [comment]
The custom Xcom works in the way that Ids are kept in the database but the actual values are kept wherever you serialize/deserialize them. If you have custom serialize/deserialize, the values are only kept where you save them.

Look at this video from Airflow Summit if you want to understand more and get more context: https://www.youtube.com/watch?v=iI0ymwOij88

**Reply by waardd** [reply]
I seen that video.
What my goal is with this, is it possible to send the XCOM key, value parirs to a central place.
Now if you push from task_id abc you must also pull from abc
But is it possible to push from abc to xyz and always pull from xyz
So all tasks can push and pull from 1 'pool'

**Reply by potiuk** [reply]
I do not understand what you mean.

**Reply by waardd** [reply]
When pushing a key,value pair it goes into a location linked to the task_id
To retrieve it one must pull the key from that task_id when in another task.
I want to know if it is possible to put all the variables produced during different tasks in one location so that you can pull from one location

For example 
Push x from task yz
Push a from task bc
Push k from task lm
And then be able to pull all of them from for instance location xx

If I want key a in task lm I have to pull from bc

Hope this helps to understand my idea

**Reply by potiuk** [reply]
Airflow XComs are attached to producer, not consumer.  They should not be used for something different. This is completely different philosophy. If you want to have multiple tasks to push to a single location you should just ... push it. Unlike the airflow DAG and dependencies go, your idea assumes that the task "knows" who is the consumer and there is no need to involve Airflow's framework for that. Unlike what is implemented in Airflow, the problem is that you don't even know by the time you produce it, whether the consumer is there at all and whether it will run. I still don't get why you want to achieve - how "pulling data from one instance" is different than "pulling same data from single instance".

Just write all your data to a single storage if you want to save it and pull it by  the consumer if you want to achieve something like that.

**Comment by waardd** [comment]
I think im getting your idea but im not sure if we are on the same page ;)
Maybe this picture explains my intention:
![image](https://user-images.githubusercontent.com/19835721/220839773-4d1a475b-5ef0-40b4-8584-a5e3cdb7d683.png)

**Reply by potiuk** [reply]
I understanda that. Simply implement your XCom backend to  push to single storage and let the otehr tasks pull from it. What prevents you from doing it now?

Airflow makes exactly "0" assumptions on where Xcom data is stored for custom backends. You are completely free to implement both pull and push the way you see fit. So if you wish to do what you described, you can do it now.

**Reply by waardd** [reply]
Okay... i have a custom xcom that i found online that writes it ti disk.
But i was under the impression that a key-value pair was only known by its task_ids.

For example is do:
ti.xcom_push(key='datamonapplog',value=datamonapplog)

and then:
datamonapplog = ti.xcom_pull(key='datamonapplog', task_ids='collect_base_variables')

If i am correct then this custom xcom would write values to disk:
import uuid
import json
from airflow.models.xcom import BaseXCom

PREFIX = 'XCOM_'

class CustomXcomFS(BaseXCom):
    """Example Custom Xcom persistence class."""

    @staticmethod
    def serialize_value(value):
        key_str = PREFIX + str(uuid.uuid4())
        with open("/tmp/"+key_str, 'w') as outfile:
            json.dump(value, outfile)
        print("Xcom with key-",key_str, " written to filesystem")
        return BaseXCom.serialize_value(key_str)

    @staticmethod
    def deserialize_value(result):
        result = BaseXCom.deserialize_value(result)
        # Check if result is string and have XCOM as prefix
        if isinstance(result, str) and result.startswitch(PREFIX):
            with open(f'/tmp/{result}', 'r') as jsonfile:
                data = json.load(jsonfile)
                return data
        return result
    
    def orm_deserialize_value(self):
        return 'XCOM writen to filesystem'

**Reply by potiuk** [reply]
Is this a question ? I am not sure wheter in the last post you are asking for something and if so, asking for what?

**Reply by waardd** [reply]
My real question is, if writing for instance to disk, would i still need to pull with specific task_ids?

**Reply by potiuk** [reply]
Yes. Producer task_ids - this is how xcom works. If you want to implement your own schema for mapping, then you have to well implement it. 

And you can also store your data in external storage in whatever way you want - you can for example write your onw data to an s3 with abstract ID. and read them with the same abstract id. and you can define the Id as you wish. 

But this has nothing to do with Airlfow DAG and how airflow handles relations between them. You are talking about your own, completely custom way of storing artifacts and retrieving them.
